compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2477,7 +2478,13 @@ impl<'a> Parser<'a> {
         let mut cond =
             self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL | Restrictions::ALLOW_LET, None)?;
 
-        CondChecker { parser: self, forbid_let_reason: None }.visit_expr(&mut cond);
+        let mut checker = CondChecker::new(self);
+        checker.visit_expr(&mut cond);
+        if checker.seen_missing_let.is_some() {
+            // Avoid unnecessary resolve errors as we know we'll have some.
+            *self.sess.silence_resolve_errors.borrow_mut() = checker.seen_missing_let;
+        }
+        self.sess.silence_missing_comparison.borrow_mut().append(&mut checker.seen_comparison);
```

Comment:
Instead of having global state here, why not just return `Err(..)`? That should halt parsing.